client/pages/study-cards.jsx

```diff
@@ -0,0 +1,232 @@
+import React from 'react';
+import LoadSpinner from '../components/loading-spinner';
+import shuffleDeck from '../lib/shuffle-deck';
+import Card from 'react-bootstrap/Card';
+import Button from 'react-bootstrap/Button';
+
+class StudyCards extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = ({
+      deck: null,
+      cards: null,
+      card: null,
+      studiedCards: [],
+      cardDrawn: false,
+      cardRevealed: false,
+      isLoading: true,
+      index: null
+    });
+
+    this.difficultyTheme = ['bg-blue', 'bg-success', 'bg-warning', 'bg-orange', 'bg-danger'];
+
+    this.drawCard = this.drawCard.bind(this);
+    this.revealCard = this.revealCard.bind(this);
+  }
+
+  drawCard() {
+    const cardsCopy = [...this.state.cards];
+
+    const currentIndex = this.state.index;
+    let newIndex = null;
+
+    if (currentIndex === cardsCopy.length) {
+      newIndex = 0;
+      const studiedCardsCopy = [...this.state.studiedCards];
+      const shuffledCards = shuffleDeck(studiedCardsCopy);
+
+      this.setState({
+        cards: shuffledCards,
+        studiedCardsCopy: null,
+        index: newIndex
+      });
+    } else {
+      newIndex = currentIndex + 1;
+      const newCard = cardsCopy[currentIndex];
+
+      this.setState({
+        card: newCard,
+        cardDrawn: true,
+        index: newIndex
+      });
+    }
+
+  }
+
+  revealCard(e) {
+    this.setState({
+      cardRevealed: true
+    });
+  }
+
+  handleRateCard(e, difficulty) {
+    this.setState({
+      isLoading: true
+    });
+    const cardId = this.state.card.cardId;
+
+    if (difficulty) {
+      const reqBody = {};
+      reqBody.difficulty = difficulty;
+
+      const req = {
+        method: 'PATCH',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(reqBody)
+      };
+
+      fetch(`/api/card/difficulty/${cardId}`, req)
+        .then(res => res.json())
+        .then(result => {
+          let studiedCardsCopy = [...this.state.studiedCards];
+          studiedCardsCopy = studiedCardsCopy.concat(result);
+          this.setState({
+            cardDrawn: false,
+            cardRevealed: false,
+            isLoading: false,
+            studiedCards: studiedCardsCopy
+          });
+        });
+    }
+  }
+
+  componentDidMount() {
+    fetch(`/api/decks/${this.props.deckId}?sortBy=difficulty`)
+      .then(res => res.json())
+      .then(data => {
+        const shuffledCards = shuffleDeck(data.cards);
+        this.setState({
+          deck: data,
+          cards: shuffledCards,
+          index: 0,
+          isLoading: false
+        });
+      })
+      .catch(err => {
+        console.error(err);
+      });
+  }
+
+  render() {
+    if (this.state.isLoading) {
+      return (
+        <LoadSpinner />
+      );
+    }
+
+    if (!this.state.cardDrawn) {
+      return this.drawCard();
+    }
```

Comment:
We never call methods with side effects in `render`. This method calls `setState`. You probably should move this into an event handlers or `componentDidUpdate` or something. Just don't call it here.